src/vocabulary/processor.rs

```diff
@@ -0,0 +1,358 @@
+                let mut chars = pattern.chars();
+                let char = chars.next();
+                if let Some(replacement) = char {
+                    if chars.next().is_none() {
+                        return Some(replacement);
+                    }
+                }
+                None
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+enum ReplacePattern {
+    String(String),
+}
+
+impl TokenProcessor {
+    /// Create new `TokenProcessor` with the level defined based on tokenizer's decoders.
+    pub(crate) fn new(tokenizer: &Tokenizer) -> Result<Self> {
+        match tokenizer.get_decoder() {
+            None => Err(TokenProcessorError::UnsupportedTokenizer),
+            Some(decoder) => match decoder {
+                DecoderWrapper::ByteLevel(_) => Ok(Self {
+                    level: TokenProcessorLevel::Byte,
+                }),
+                DecoderWrapper::Sequence(decoding_sequence) => {
+                    let mut is_byte_fallback = false;
+                    let mut spacechar = ' ';
+
+                    for decoder in decoding_sequence.get_decoders() {
+                        match decoder {
+                            DecoderWrapper::ByteFallback(_) => {
+                                is_byte_fallback = true;
+                            }
+                            DecoderWrapper::Replace(replace) => {
+                                // `Replace` decoder would replace a pattern in the output with something else,
+                                // which we need to know.
+                                let decoder = Self::unpack_decoder(replace)?;
+                                if let Some(replacement) = decoder.space_replacement() {
+                                    spacechar = replacement;
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+
+                    if is_byte_fallback {
+                        Ok(Self {
+                            level: TokenProcessorLevel::ByteFallback(Mods { spacechar }),
+                        })
+                    } else {
+                        Err(TokenProcessorError::UnsupportedTokenizer)
+                    }
+                }
+                _ => Err(TokenProcessorError::UnsupportedTokenizer),
+            },
+        }
+    }
+
+    /// Operates on each token based on the level of `TokenProcessor`.
+    pub(crate) fn process(&self, token: String) -> Result<Vec<u8>> {
+        match &self.level {
+            TokenProcessorLevel::Byte => {
+                let mut bytes = vec![];
+                for char in token.chars() {
+                    match CHAR_MAP.get(&char) {
+                        None => return Err(TokenProcessorError::ByteProcessorFailed),
+                        Some(b) => bytes.push(*b),
+                    }
+                }
+                Ok(bytes)
```

Comment:
Otherwise, if this isn't a performance bottleneck, we could keep the new buffer, but avoid the mutation and control flow, with something like this:

```suggestion
                let bytes = token
                  .chars()
                  .iter()
                  .map(|c| CHAR_MAP.get(c)
                    .map_or(Err(TokenProcessorError::ByteProcessorFailed), Ok))
                  .collect()?;
```

Reply:
Ah, sorry, I mentioned performance not because I thought my suggestion is slower than what you had, but because it's mutually exclusive with my other comment (`.clear()`ing, and reusing the `bytes` buffer).

Although now I come to think of it, they're not really mutually exclusive at all, because you can just use `bytes.extend(<functional stuff>)`